0023-communication-protocol.adoc

```diff
@@ -0,0 +1,388 @@
+}
+----
+1. The swap-id from the `/comit/swap/announce/1.0.0/` protocol
+2. An Ethereum address
+
+Protocol identifier: `/comit/swap/identity/lightning/1.0.0`
+
+[source,json]
+----
+{
+  "swap_id": "713b0c93-53a1-46b2-9418-de8d46750b79", // <1>
+  "pubkey": "0236055ab508ec9495c07a3968f7ee9a1b5983f70e24d362f677d2619419601052" // <2>
+}
+----
+1. The swap-id from the `/comit/swap/announce/1.0.0/` protocol
+2. A Lightning node public key (taken from `lncli getinfo`)
+
+==== Why do we not need more protocols?
+
+Given the context of a swap, all of these protocols carry unambiguous information.
+Consider the following example:
+
+Assume `Bitcoin` is the alpha-ledger and `Ethereum` is the beta-ledger.
+In said constellation, Alice's refund identity is a Bitcoin public key and here redeem identity is an Ethereum address.
+
+If Alice would send a message using the `/comit/swap/identity/bitcoin/1.0.0` protocol, it is unambiguous for Bob that Alice is sending her refund identity.
+
+More generally speaking:
+Given knowledge of the alpha and beta ledger and the role, a message can be unambiguously identified as the redeem or refund identity of an actor depending on whether the message is incoming or outgoing.
+
+This allows us to slim down the protocols to the bare minimum of sending over an identity.
+
+=== Finalizing a swap
+
+One of the downsides of splitting the communication for a single swap up into several protocols is the fact that we don't have a natural "end" of the communication.
+If all of the communication were to take place within a single protocol, closing the substream could act as "I am done communicating".
+
+Following a multiple protocol approach, the answer to how we communicate this concern is another protocol.
+
+Introducing: `/comit/swap/finalize/1.0.0`
+
+[source,json]
+----
+{
+  "swap_id": "713b0c93-53a1-46b2-9418-de8d46750b79"
+}
+----
+
+Both parties are expected to use this protocol after they have sent all the data they consider necessary for starting the execution of the swap.
+Once each party sent a finalize message AND received one, implementations should consider the swap finalized and start with the execution phase (i.e. watch for transactions, emit actions, etc)
+
+== Discussion points
+
+This section includes discussion topics that remain to be answered.
+
+=== Should we use multihash for the swap digest
+
+<<Announce-message,Section 3.2>> introduces the concept of a swap digest.
+In order to be forwards compatible with different hash algorithms, it proposes to encode the merkle root as a multihash.
+That way, the hash is self-describing and the receiving party knows, which hash algorithm needs to be used to compute the digest.
+
+Do we need the flexibility of changing the hash function that is used to compute the swap digest without a breaking change to the protocol?
+We could always just create `/comit/swap/announce/2.0.0` that defines a different hash function.
```

Comment:
It is a difficult question. On one hand I believe that the hash algorithm should be treated like the expiries. It can be "defaulted" and hence not present in the merkle tree or always present with the same value.

By using a multihash we are taking a characteristic of the swap and choose/specify it in a separate place than the rest of the characteristics (amounts, expiries).

Moreover, it means that we need to have added error handling logic there. I think it should all be in the Merkle tree.

Reply:
I am talking about the swap digest here, not the hash algorithm of the HTLC.